src/NLog/Config/LoggingConfigurationParser.cs

```diff
@@ -184,78 +184,61 @@ private void SetNLogElementSettings(Dictionary<string, string> dict, out bool? p
-        private void SetThrowConfigExceptions(IDictionary<string, string> dict)
-        {
-            if (dict.TryGetValue("THROWCONFIGEXCEPTIONS", out var val))
+            if (!internalLoggerEnabled && !InternalLogger.HasActiveLoggers())
             {
-                LogFactory.ThrowConfigExceptions = StringHelpers.IsNullOrWhiteSpace(val)
-                    ? (bool?)null
-                    : ParseBooleanValue("ThrowConfigExceptions", val, false);
+                InternalLogger.LogLevel = LogLevel.Off; // Reduce overhead of the InternalLogger when not configured
             }
-        }
 
-        /// <summary>
-        /// Set <see cref="LogFactory.ThrowExceptions"/>
-        /// </summary>
-        /// <param name="dict"></param>
-        private void SetThrowExceptions(IDictionary<string, string> dict)
-        {
-            if (dict.TryGetValue("THROWEXCEPTIONS", out var val))
-            {
-                LogFactory.ThrowExceptions = ParseBooleanValue("ThrowExceptions", val, LogFactory.ThrowExceptions);
-            }
+            _configurationItemFactory = ConfigurationItemFactory.Default;
+            _configurationItemFactory.ParseMessageTemplates = parseMessageTemplates;
         }
 
         /// <summary>
-        /// build dictionary, use last value of duplicates
+        /// Builds list with unique keys, using last value of duplicates. High priority keys placed first.
         /// </summary>
         /// <param name="nlogConfig"></param>
         /// <returns></returns>
-        private static Dictionary<string, string> CreateNLogConfigDictionary(ILoggingConfigurationElement nlogConfig)
+        private static List<KeyValuePair<string,string>> CreateUniqueSortedListFromConfig(ILoggingConfigurationElement nlogConfig)
         {
             var dict = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
             foreach (var configItem in nlogConfig.Values)
             {
-                if (configItem.Key != null)
+                if (!string.IsNullOrEmpty(configItem.Key))
                 {
                     dict[configItem.Key.Trim()] = configItem.Value;
                 }
             }
 
-            return dict;
+            var sortedDict = new List<KeyValuePair<string, string>>(dict.Count);
+            AddHighPrioritySetting("ThrowExceptions", dict, sortedDict);
+            AddHighPrioritySetting("ThrowConfigExceptions", dict, sortedDict);
+            AddHighPrioritySetting("InternalLogLevel", dict, sortedDict);
+            AddHighPrioritySetting("InternalLogFile", dict, sortedDict);
+            AddHighPrioritySetting("InternalLogToConsole", dict, sortedDict);
+            foreach (var configItem in dict)
+            {
+                sortedDict.Add(configItem);
+            }
+            return sortedDict;
+        }
+
+        private static void AddHighPrioritySetting(string settingName, Dictionary<string, string> dict, List<KeyValuePair<string, string>> sortedDict)
```

Comment:
Could we setup a type that encapsulates the list and dictionary? That type will be responsible for keeping things in sync.

Reply:
Not sure I understand the dictionary is temporary and is thrown away. The List is the final result.

Reply:
Hard to explain. I could refactor it when it's done (as PR)

Reply:
moved it to an inner method. To encapsulate the AddHighPrioritySetting. That's good enough for now.

(we need some refactor change in the future, as it's strange that we build up a dictionary, change it to list and then use a switch. I think the loop over the switch should be removed and replaced with direct dictionary access (so you can set the ordering easier)

Reply:
I read your statement `type that encapsulates the list and dictionary` as creating some weird new container. But yes inner-method helps here, though they ruin readability especially on organic growth beyond 3-4 lines. I guess a better solution would be a Linq-expression or yield that returns IEnumerable.

Reply:
> as creating some weird new container.

True, therefor I also skipped that. 

I think passing parameters to a method which should be in-sync (sorted list - dictionary) is a code smell and I fixed that now with the inner method. Inner methods should be avoided, but for this case it's perfect IMO (or at least, better then the previous situation and the wierd container). 

Another option was to create a new enumerator, but that's overkill for one situation I guess.